### PR TITLE
refactor: split into client, cli and worker packages

### PR DESCRIPTION
**Pull request by lenaschoenburg**
Apologies for the large PR, feel free to treat this as a prototype if you'd like.

This splits up this repository into multiple crates, `client` which contains the generated gateway api client, `cli` which contains a tool for testing the client by implementing zbctl like functionality, and `worker` which is completely empty now but will contain code that supports defining workers in rust.

The PR also removes some of the experiments that re-defined gateway api types and mapping code between from auto generated protobuf types. I've found that the auto-generated types are usable as-is.

### CODE REVIEW

File: cli/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        ),
+    };
+
+    println!("{:#?}", response);
+
+    Ok(())
+}
+
+fn deploy_request(args: DeployArgs) -> Result<DeployResourceRequest> {
+    let mut resoures = Vec::with_capacity(args.resources.len());
```
Comment (pihme) [inline cli/src/main.rs]: :x:
```suggestion
    let mut resources = Vec::with_capacity(args.resources.len());
```

File: cli/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                .into_inner(),
+        ),
+    };
+
+    println!("{:#?}", response);
+
+    Ok(())
+}
+
+fn deploy_request(args: DeployArgs) -> Result<DeployResourceRequest> {
```
Comment (pihme) [inline cli/src/main.rs]: :wrench:
```suggestion
fn buld/create_deploy_request(args: DeployArgs) -> Result<DeployResourceRequest> {
```

**Review by pihme — CHANGES_REQUESTED** [review]
Thanks for contributing to this project.

I am fully on board with splitting it up in multiple crates.

I am a little on the fence regarding deleting the existing mapping code. I agree that right now it is mostly over-engineering and doesn't add too much value. I guess the main purpose was to separate the API a user of the Rust library sees from the API that we use internally. 

This is also what we do in the Java client (e.g. `ActivatedJobImpl`). One future benefit might be that we could offer - within limitations - a stable Rust API by remapping requests and responses to a changing Proto API. 

At the same time, this is more a theoretical benefit. And even in the Java Client I feel it borders on over-engineering. 

So that's why I am on the fence about it. So you can be the one to tip the scale. 

If you think there is no merit, let's just delete it right away. 

If you think there is something to it, let's park it in another crate like _experimental_  and let's see whether we want to maintain this crate, or we will abandon it as it slows us down.

**Comment by lenaschoenburg** [comment]
Thanks @pihme!

I'd like to tip the scales towards not having the mapping code. IMO the protobuf API is changing slow enough (only every 6 months) that it's not worth it to keep. Also, we can actually add it later, like this:

Define a private module that includes the generated types:
```rust
mod generated_api {
    tonic::include_proto!("gateway_protocol");
}
```

Re-export them publicly:

```rust
pub mod api {
    pub use super::generated_api::*;
}
```

Now, if we make incompatible changes to TopologyRequest, we can redefine the type and implement a mapping:

```rust
pub mod api {
    pub use super::generated_api::*;

    #[derive(Clone, PartialEq, Eq, ::prost::Message)]
    pub struct TopologyRequest {}

    impl tonic::IntoRequest<super::generated_api::TopologyRequest> for TopologyRequest {
        fn into_request(self) -> tonic::Request<super::generated_api::TopologyRequest> {
            todo!()
        }
    }
}
```

That means we can leave room to keep backwards compatibility without paying any of the costs up front :tada: 

I've added a commit that does the re-exporting of generated types. Does that sound good to you?

**Review by pihme — APPROVED** [review]
Looks excellent!

I think you found a great solution to future proof this.

**Pull request merged**